### ISSUE TITLE
Parallelism woes in pattern matcher

### ISSUE THREAD

**Issue by linas** [opencog/atomspace#2758]
In principle, the pattern matcher can run in an embarrassingly-parallel mode. In practice, it does not and I think I finally understand why: atomic compare-n-swap contention in `std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>` where the `2` means `_S_atomic` in `/usr/include/c++/8/ext/concurrence.h:  enum _Lock_policy { _S_single, _S_mutex, _S_atomic };`  That sounds arcane so here's the long explanation, and it points to a way to fix the problem.

During graph traversal, current state needs to be maintained: what graph atoms have been visited, what matches have been found, etc. This state us usually kept in `std::Map<Handle,Handle>` (there are several variants of this in use) When there are different branches to explore, the current state has to be pushed onto a stack, and popped during back-tracking. Like so: `std::stack<std::Map<Handle,Handle>>` And that's the problem, right there. 

So, `Handle` is a smart pointer, an `std::shared_ptr<Atom>`, and each new Handle increments the use-count on `Atom`. The use count is stored in `std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>`  and each increment uses compare-n-swap.  Each time a Handle is placed into `std::Map<Handle,Handle>` the increment is performed. Each time current state is pushed onto  `std::stack<std::Map<Handle,Handle>>`, there may be hundreds (or thousands?!) of increments of use-counts in rapid succession. Pattern matching is mostly so fast and efficient, that a large part of the time is spent pushing and popping, backtracking. .. There's just not much to do during pattern matching - just crawl, check, and go forward, or backtrack. 

Running 8 or 16 threads, each doing "independent" pattern searches will often explore a common set of atoms, and thus will be incrementing and decrementing use-counts like crazy. They'll be hitting compare-n-swap like crazy, and since compare-n-swap lives on cache lines, this forces a huge amount of cache-line contention and cache snoops and cache flushes. This touches the cache lines for *all* CPU's and not just one, since the atomic might be anywhere and all cpu caches have to be snooped and maybe invalidated. ... This interferes with parallelism.  Basically, the pattern matcher is whacking the cache with vast numbers of compare-n-swap.  ... at least I think that is what's happening. I'm pretty sure.

The solution is to avoid the increments/decrements. Thus, the pattern matcher needs to hold state in `std::Map<const Handle&, const Handle&>` (equivalently, in `std::Map<Handle *, Handle *>`)  In order for this to be safe, we have to store a single copy of a Handle somewhere ... in some safe place, so the Handle won''t go away as long as a pattern match is happening.  For example, in a `std::Set<Handle>`. Once its safely kept, then we can use `Handle*` in the `std::map`.

Doing this is not conceptually hard .. but it is incredibly tedious. and practically hard, because it has to touch lots of code, in lots of places. So .. Yuck .. its a big project :-(  But should be a big win for parallelism.

FYI, this was discovered in performance analysis on @Habush 's https://github.com/MOZI-AI/annotation-scheme/ -- the old code, before he re-wrote all of it to make grpc calls. (So, @Habush  I assume you still do pattern matching .. where is the pattern matching code now, seeing how its been ripped  out of annotation -scheme? It must be in some new location? Where? Does it have the same API as before, or did the API change? What about all the work that went into tuning the old version? Was it discarded? How are you tracking performance of this thing?  Where are your performance benchmarks and benchmark reports? Do you update them regularly? .. etc.. inquiring minds want to know...)

**Comment by Habush** [comment]
> The solution is to avoid the increments/decrements. Thus, the pattern matcher needs to hold state in std::Map<const Handle&, const Handle&> (equivalently, in std::Map<Handle *, Handle *>)

How about using a `std::weak_ptr`? i.e `std::Map<HandleWkptr, HandleWkptr>` where `HandleWkptr` is `std::weak_ptr<Atom>`

> So, @Habush I assume you still do pattern matching .. where is the pattern matching code now, seeing how its been ripped out of annotation -scheme? It must be in some new location? Where? 

The pattern matching code has moved to the grpc server. To specific, it is [here](https://github.com/Habush/atomspace-rpc/blob/c49b994f5d6fbe798ad86ceafede7f501696010c/src/manager/AtomSpaceManager.cpp#L67). 

> Does it have the same API as before, or did the API change? What about all the work that went into tuning the old version? Was it discarded?

No, it wasn't discarded. I just moved where the pattern matching search happens. We still use caches in the local(client) atomspace, i.e if the result exists in a cache, the client won't send an rpc call and just returns the cached result. The interface to the scheme functions to access the pattern matcher hasn't changed. If you remember, you added a function named  [run-query](https://github.com/MOZI-AI/annotation-schem/blob/1698fff6a1b80204d4d8db439946978b15b7a5b0/annotation/util.scm#L108) where it takes an atomese code representing a pattern search and runs `cog-execute!` . What I did is just replace the `cog-execute!` in `run-query` with `exec-pattern("atom-id" QUERY)`, which sends `QUERY` to the server and executes it in an atomspace with id `atom-id`.

> How are you tracking performance of this thing? Where are your performance benchmarks and benchmark reports?

I use [ghz](https://github.com/bojand/ghz) to load test the server. Although the grpc server is multit-hreaded, sending concurrent request doesn't perform well, which could be due to the pattern matching issue you just described above. 
You can find the latest benchmark results [here](https://mozi.ai/benchmarks/rpc-benchmark-2020-08-19.html). In this benchmark, I am sending 100 requests using 10 threads and 10 connections. The pattern matching query I am sending is the one we use to find gene interactors in a pathway.(here is the [function](https://github.com/MOZI-AI/annotation-scheme/blob/1698fff6a1b80204d4d8db439946978b15b7a5b0/annotation/functions.scm#L763))

**Comment by linas** [comment]
> std::weak_ptr?

Weak pointers still increment a use-count atomically -- they just increment the weak-use-count, not the strong-use-count. So: the referenced object can be deleted when the strong use-count drops to zero. However, the place where use-counts are stored can only be deleted when the weak-use-count drops to zero. 

> has moved

I sort-of would have been happier if this had been "layered", so that one could use either the old code, directly, or access it via grpc... that would have been cleaner, easier for many things... but whatever.

> exec-pattern("atom-id" QUERY)

FYI, I invented almost the same thing, but slightly differently. It's a two-parter. First, there is now a new function `cog-execute-cache!` which runs `cog-execute!` but places the result at a user-specified key. Called a second time, it will return the cached value, instead of re-running again. Optionally, it stores meta-data, i.e. the datestamp of the result. You can clear the cache simply by deleting the cache-key. This works for anything executable, not just queries. For docs say `,describe cog-execute-cache!` at the guile prompt.

Part two is the new `fetch-query`, which is "just like"  `cog-execute-cache!`  except that (1) it runs in the backend and (2) is restricted to work only for queries. This works with any backend - the sql backend, the file backend, and the cogserver backend. When used with the cogserver backend, it does more-or-less the same thing as your `exec-pattern("atom-id" QUERY)`

> benchmark

Thank you, I will study. Also, FYI, see https://github.com/opencog/benchmark/tree/master/query-loop for a snapshot of your (old) datasets and (old) code, morphed into a benchmark for the pattern matcher. 

Also FYI See https://wiki.opencog.org/w/ExecuteThreadedLink which can run multiple queries in parallel. So, parallelism in Atomese.

> profiled 

Profiling is a real pain-in-the-neck, because it fails to show the cache-line contention I describe above. I have not yet figured out how to use `perf` to confirm this.  But yes, pretty much anything you do will spend 99% of it's time in `opencog::MeetLink::do_execute` which is why (1) you want to design your queries in the most clever way possible, (2) use `cog-execute-cache!` or anything else like that, to not repeat old searches, (3) threading is best, because `opencog::MeetLink::do_execute` should have been "embarassingly parallel". ... it might still be fore you; I'm not sure how much of the problem I'm seeing is due to the `query-loop` benchmark, and how much of the problem is generic.

Oh, and (4) The new rocks-db backend might be fast enough for you to load the genome data, pre-compute a whole giant mess of searches, (taking days or how-ever-much CPU time), save the results to the RocksDB backend.  Then, next time you restart the grpc server, just connect to the rocksdb, and use the new `fetch-query` to get the saved query results.  That way, every user gets "instant" results, because the query was run some-time in the past, and was saved to disk.

**Comment by linas** [comment]
The parallel pattern matcher can be enabled by editing `opencog/query/CMakeLists.txt` and uncommenting
```
TARGET_COMPILE_OPTIONS(query-engine PRIVATE -DUSE_THREADED_PATTERN_ENGINE=1)
```
and recompiling, installing.  I think the result passes unit tests, I'm not sure.

**Comment by linas** [comment]
Closing. I just remeasured `opencog/benchmark/query-loop/bio-loop3.scm` on a modern CPU. It parallelizes very nicely up to 8 threads. (This is a 4-core CPU with 8 hyperthreads: an AMD Ryzen 5 3400G)

Basically, all claims made above are correct. Actually parallelizability is bottlenecked on the CPU implementation of atomic increment. To recap:

The AtomSpace uses `std::shared_ptr` for everything, which uses atomic increment for maintaining use counts. The code in the AtomSpace has a vast amount of contention for this increment (even if the increment is occurring on unrelated items: they still contend for the CPU atomics.) Modern CPU's handle this nicely; old CPU's do not! 

All of my measurements were on an old CPU, which bottlenecks at 4 threads, despite having many more cores. Remeasuring on a new CPU shows .. there is no problem!
